### PR TITLE
Train parser on actual wikitables data

### PR DESCRIPTION
**Pull request by matt-gardner**
@pdasigi, FYI.  This PR is tracking progress on actually using real data with our parser.  I've run into a number of inconsistencies in how sempre produces logical forms versus how we produce them, and issues where we weren't generating a full action space (and a few corner cases where we still aren't, but that's probably fine - it's typically where DPD had bad output in the first place).  Fixing the issues has gotten me a lot more familiar with the data / LF processing code, though, which is a good thing.

This depends on #606, so the diff here is way bigger than it should be.  Just look at the commits starting with `fffa5bf` to see the changes specific to this PR.

### CODE REVIEW

**Comment by matt-gardner** [comment]
I'm now able to train on the full data, though there are still a few unicode processing corner cases, and places where we throw out DPD logical forms because they use numbers they shouldn't.  This PR is basically ready as-is, but it's failing CI because of some issues that'll be fixed on a rebase.  After merging #606 (and possibly also your NLVR PR), I'll rebase the wikitables branch, then fix this PR.

**Comment by matt-gardner** [comment]
Ok, I rebased this, so the diff is now showing up correctly, and it's ready for review @pdasigi.

This PR mostly fixes parsing errors that I ran into when running on the full data, and adds some error handling and logging for the remaining cases when parsing fails.  I think the majority of changes here are just new tests to catch the errors I ran into.

I made the parser fail if it tried to parse a name that wasn't already in the world's context.  I think this is an improvement, as it catches errors earlier.  But it meant I removed your "large-scale processing" test, and its associated file.

I  also added a parse accuracy metric (thanks @OyvindTafjord!), and I fixed a bug in how batching works.  Oddly, though, batching is now _significantly_ slower than running with a batch size of 1.  My guess is that this is because of all of the terminal actions making the matrix lookups in the global actions array take a whole lot longer, but I haven't profiled this or anything yet.

Running time on the full data: reading in the data takes ~20-25 minutes on my EC2 machine, where the bulk of the time is spent in logical form parsing.  We need to figure out how to make that faster.  Because I only had DPD output for about 60% of the training data, I ended up with about 6k training instances, and training took about 15 minutes per epoch with a batch size of 1, and 10 logical forms per instance.  It didn't actually learn anything, because there are important pieces of the model still missing (the entity linking bits).

File: allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py [hunk 1]
```diff
@@ -168,20 +203,45 @@ def text_to_instance(self,  # type: ignore
...
             # ProductionRuleFields.
             action_map = {action.rule: i for i, action in enumerate(action_field.field_list)}  # type: ignore
+
             action_sequence_fields: List[Field] = []
-            for sequence in action_sequences:
-                index_fields: List[Field] = []
-                for production_rule in sequence:
-                    index_fields.append(IndexField(action_map[production_rule], action_field))
-                action_sequence_fields.append(ListField(index_fields))
+            for logical_form in dpd_output[:self._max_dpd_tries]:
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py]: Is the docstring that's there not good enough? https://github.com/allenai/allennlp/pull/639/files/b50689afe6bcd2fc2953b9e53ad61428f9f03708#diff-477d5e4ffb0f9e5c0ad722b13e6082e0R68

What would you change?

File: allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py [hunk 2]
```diff
@@ -214,16 +274,37 @@ def _parse_line_as_lisp(lisp_string: str) -> Dict[str, Union[str, List[str], Non
...
+        # DPD has funny ideas about long strings of "ors" being reasonable logical forms.  They
+        # aren't, and they crash our recursive type inference code.  TODO(mattg): we need to fix
+        # the type inference code to not die in those cases, somehow...
+        if logical_form.count('(or') > 3:
+            logger.debug(f'Skipping logical form with inordinate number of "ors": {logical_form}')
+            return False
+        if 'fb:part' in logical_form:
+            # TODO(mattg): we don't currently ever create production rules to generate cell parts,
+            # and it's not clear to me why we ever should.  These always fail to parse right now,
+            # so we'll just skip them and fix it later.
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py]: :+1:

And yes, this is a very small number of logical forms that actually use this.

File: allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py [hunk 3]
```diff
@@ -60,14 +57,34 @@ def read_from_json(cls, json_object: Dict[str, Any]) -> 'TableKnowledgeGraph':
...
         neighbors: DefaultDict[str, List[str]] = defaultdict(list)
         # Following Sempre's convention for naming columns.
-        columns = ["fb:row.row.%s" % cls._normalize_string(x) for x in json_object["columns"]]
+        column_ids = []
+        columns: Dict[str, int] = {}
+        for column_string in json_object['columns']:
+            normalized_string = f'fb:row.row.{cls._normalize_string(column_string)}'
+            if normalized_string in columns:
+                columns[normalized_string] += 1
+                normalized_string = f'{normalized_string}_{columns[normalized_string]}'
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: Yes, for columns it is.  That's how I found this.

File: allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py [hunk 4]
```diff
@@ -60,14 +57,34 @@ def read_from_json(cls, json_object: Dict[str, Any]) -> 'TableKnowledgeGraph':
...
+            columns[normalized_string] = 1
+            column_ids.append(normalized_string)
+
+        # TODO(mattg): I know that sempre gives separate ids to columns that collide on
+        # normalization; I don't know if sempre similarly does this for cells that collide.  We'll
+        # implement it anyway.
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: If sempre doesn't do it, then you're right, we shouldn't do it.  Ignoring sempre for a minute, though, why should we have cell identifiers that are not unique?

File: allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py [hunk 5]
```diff
@@ -93,15 +110,20 @@ def _normalize_string(string: str) -> str:
...
+        # need for this function somehow?  It's causing a whole lot of headaches.
+        string = re.sub("[ðø]", "_", string)
+        string = string.replace("\\n", "_")
         string = re.sub("\\s+", " ", string)
         # Canonicalization rules from Sempre
         string = re.sub("[^\\w]", "_", string)
         string = re.sub("_+", "_", string)
         string = re.sub("_$", "", string)
-        return string.lower()
+        return unidecode(string.lower())
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: All of the changes that I made to this function (and there are more coming in later PRs) are to match the normalization that's in the DPD output.  So, there are apparently some normalization operations done by sempre that were not captured in the current code.

File: allennlp/data/semparse/worlds/wikitables_world.py [hunk 6]
```diff
@@ -56,15 +57,30 @@ def _get_numbers_from_tokens(self, tokens: List[Token]) -> List[str]:
         Eventually, we'd want this to detect things like ordinals ("first", "third") and cardinals
         ("one", "two"), but for now we just look for literal digits, and make up for missing
         ordinals and cardinals by adding all single-digit numbers as possible numbers to output.
+
+        We also handle year ranges expressed as decade or centuries ("1800s" or "1950s"), adding
+        the endpoints of the range as possible numbers to generate.
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/semparse/worlds/wikitables_world.py]: There is one: https://github.com/allenai/allennlp/pull/639/files/b50689afe6bcd2fc2953b9e53ad61428f9f03708#diff-712f4cefb411e7b366723927e25e53b5R63.  The issue is that we want to allow generating numbers when we're in a state that outputs that kind of entity.  But we don't want to have an infinite set of numbers that we can output.  Which numbers should we allow?  What I've implemented is that you can generate numbers that are evoked by the question text.  Sempre does this too, but also allows the LF to contain numbers evoked by table cells independent of the question - I didn't see any reason to do that, given that we're parsing questions, so I didn't implement it.  But if we see "1800s" in the question, we typically need to parse that into something like "x > 1800 && x < 1900".

File: tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py [hunk 7]
```diff
@@ -16,15 +16,96 @@ def test_reader_reads(self):
...
+                'fb:cell.6_851',
+                'fb:cell.7_169',
+                'fb:cell.8_567',
+                'fb:cell.9_734',
+                'fb:cell.did_not_qualify',
+                'fb:cell.did_not_qualify_2',
+                'fb:cell.did_not_qualify_3',
+                'fb:cell.did_not_qualify_4',
+                'fb:cell.did_not_qualify_5',
+                'fb:cell.did_not_qualify_6',
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py]: Ok, I'll change it back.  I've also since realized that sempre doesn't do this, after looking at a bunch more data.  I'm not not convinced it's correct, though...

File: tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py [hunk 8]
```diff
@@ -113,6 +194,7 @@ def test_reader_reads(self):
                 "<e,r> -> fb:row.row.avg_attendance",
                 "<e,r> -> fb:row.row.division",
                 "<e,r> -> fb:row.row.league",
+                "<e,r> -> fb:row.row.null",  # null row, representing an empty set
```
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py]: https://github.com/allenai/allennlp/pull/639/files/b50689afe6bcd2fc2953b9e53ad61428f9f03708#diff-e6b5c1785059fb13cf2030a09d3daf2eR44.

I'll fix the comment, here and in the place I linked.

File: allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py]: I would change "We will try at most `max_dpd_tries` logical forms before giving up." -> "We will try to parse only the first `max_dpd_tries` logical forms."
To me, "try at most ... before giving up" implied trying to parse the same logical form several times before giving up.

File: allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: Continuing with the same example as above, say we have a question on that table like "How many movies were released in 2012?", then we expect the logical form to be something like `(count (fb:row.row.year fb:cell.2012))`, for which to give the right answer, all the cells that contain the value "2012" should have the same identifier. Why would you want to produce `fb:cell.2012_2`, `fb:cell.2012_3` etc.?

File: allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: I see. The original set of rules in this method were taken from Sempre's code directly. But I guess Sempre's normalization logic is not all in one place.

File: allennlp/data/semparse/worlds/wikitables_world.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/semparse/worlds/wikitables_world.py]: Great! That makes sense.

File: tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pdasigi) [reply tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py]: "Did not qualify" occurring in multiple rows is still the same value, right?

File: tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (pdasigi) [reply tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py]: Thanks for the pointer. I missed that part earlier. Can you please also explain why you're generating null cells and columns? I didn't get that yet.

File: allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: Yes, that makes a lot of sense.  But also consider the table:

| Movie | Year |
| --- | --- |
| 1 | 2012 |
| 2 | 2012 |
| 3 | 2012 |

What are the neighbors of the "year" cell?  How many are there?  With our current representation it sure looks like there is one neighbor.  I think there's value in being able to both refer to a cell by some unique identifier, and query for all cells with a given text.

My issue was that I was thinking of `fb:cell.x` as the identifier for a particular cell (as that's pretty much implied by how our knowledge graph code works).  But that's not how sempre defines `fb:cell.x` - it's a function that returns all cells with text `x`.  That's fine, and such a function is useful, it's just odd to be defining table entities that way.

[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: In fact, it makes me want to change the type of `fb:cell.x`...  But we can worry about that later.

File: allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py]: Done.

File: allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: Or python handles unicode differently from java...

File: tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py]: It is now.  See my comment above.  I've put this back to how it was, only detecting duplicates for columns.

File: tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (matt-gardner) [reply tests/data/dataset_readers/wikitables/wikitables_dataset_reader_test.py]: Because DPD / sempre produce them.  If I'm remembering right, it's for places where you're checking that something is not an empty set (imagine, "is there a team that won three games in a row?").

File: allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: Ah, yes. I see your point and I agree that table entities should be defined differently. We should separate the knowledge graph representation from logical form representation to avoid confusion.

[the remark this answers is not in the document]
Comment (matt-gardner) [reply allennlp/data/semparse/knowledge_graphs/table_knowledge_graph.py]: I think for now we should just do what sempre does; we can think about doing something smarter when we write a code API for this, and ditch sempre / DPD entirely.

**Comment by matt-gardner** [comment]
@pdasigi, I've pushed the changes you mentioned; is this good to merge?

File: allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/dataset_readers/wikitables/wikitables_dataset_reader.py]: Since it's more likely that longer logical forms are bad, you may want to sort the logical forms by length before pruning them here.

**Comment by matt-gardner** [comment]
That actually slows down the dataset reading quite a bit; we can dramatically speed up load times by only looking at the first `k` lines.  Do you know how DPD sorted them in the first place?

**Comment by pdasigi** [comment]
Not sure how DPD sorts them, but I know it's not by length. We can sort them offline, though.

**Comment by matt-gardner** [comment]
There are also a whole bunch of files missing (about 1/3 of the data) - do you know why, or do you have files for the missing instances?
